### DISCUSSION TITLE
AAPS - xDrip synchronization

### DISCUSSION THREAD

**Discussion by MilosKozak** [NightscoutFoundation/xDrip#2636] [Ideas]
Hi @jamorham,
I'm cleaning up xDrip broadcasting code in AAPS. During years it became a mess. My current observation is:

* there is Status Line broadcasted via "com.eveningoutpost.dexdrip.ExternalStatusline"
* calibration via "com.eveningoutpost.dexdrip.NewCalibration"
* BGs via "com.eveningoutpost.dexdrip.NS_EMULATOR" if collected from other source than xDrip

NS callback is forwarded to xdrip: 
* BGs via "info.nightscout.client.NEW_SGV"
* treatments via "info.nightscout.client.NEW_TREATMENT"

It seems xDrip code "info.nightscout.client.*" is accepted in Follower mode only which means if people are using NS_EMULATOR (i think most used case) they must download insulin and carbs from NS anyway

I prepared new plugin to AAPS whre i can effectively broadcast changed/updated data. Is there any will to define an API for accepting data from AAPS?

Milos

**Comment by olorinmaia** [comment]
My main wish is if you broadcast treatments, basal adjustments and profiles to xDrip is that it makes the predicted low smarter. As it stands now for me it doesnt seem to have any effect on the prediction.

Secondly, if treatments are deleted in AAPS (example carb entry adjusted, first deleted then added correct amount) it would be nice if such changes also were made in xDrip. Now it adds treatments, if u delete in AAPS, it doesnt remove it in xDrip. This will in turn make predicted low better.

**Comment by jamorham** [comment]
@MilosKozak very happy to work with you to achieve whatever is needed for more data to be transferred. I am not aware that follower mode needs to be used. The main problem that I know of so far is that unless nsclient is enabled that AAPS doesn't broadcast to xDrip.

**Reply by MilosKozak** [reply]
where can we chat?

**Reply by MilosKozak** [reply]
@jamorham

**Reply by Navid200** [reply]
@MilosKozak You can communicate here unless you need to disclose something that is meant to be kept private.

**Comment by MilosKozak** [comment]
- [ ] Topic 1)
What kind of data is usefull for xDrip except BGs, Insulin, Carbs?
- [ ] Topic 2)
How to integrate reciever best way (see my observations above)?
- [ ] Topic 3)
What is preffered data format (JSON NS v1, JSON NS v3, Java object ....)? 
- [ ] Topic 4) 
Do we still need status line?

**Reply by jamorham** [reply]
1)
I think it is possible for almost everything to already be sent over but only if nsclient is enabled. What would be good is for this to be independent of nsclient. In the fork of AAPS I run locally, I made it so that where things would be added to nsclient instead they are just broadcast but I think this is the same practical effect as if nsclient was running and enabled. A better and official way to achieve this would be much appreciated.

From memory the only things I can think of that are missing would be things like insulin remaining in pump (all pump status basically) and solid information about TBR etc as currently we are parsing the status line and not getting the data from a firm data set.

xDrip imports basal profile changes for its internal representation but also has its own graphical basal profile editor. It would be good if it were possible to send basal profile changes back to AAPS but that is perhaps beyond the scope of what you are asking. Additionally I have support for controlling boluses from xDrip via watch integration but this is a very rough implementation. Some kind of secure controlled way of implementing something like this would also be good but this is also potentially outside the discussed scope here.

2)
There is a lot of support for AAPS stuff via `NSClientReceiver` where mostly this is handled by processing treatments for boluses, pump events etc. This interface should be very easily expandable for any new types of data. Likewise xDrip can emit any additional status messages you would be interested in receiving in AAPS.

3) JSON works well and is portable in terms of if the structure changes a bit that it is tolerant of that.

4) Status line is still actively being used. Indeed I think it is 100% of how we know AAPS IoB and TBR which is essential for displaying the basal graph in xDrip and forwarding the same to watches etc.

There is probably more but this is off the top of my head.

**Comment by MilosKozak** [comment]
1) I have the code ready (independent to nsclient) as another sync module. I can send devicestatus (the same as sent to NS) which have all actual info. It can replace statusline and it will be in readable form without parsing needed
2) can you update code in `NSClientReceiver` to not be depended on follower mode?
3) there is ns-sdk module for NS api v3 integrated in AAPS ( https://github.com/nightscout/AndroidAPS/tree/xdrip/core/ns-sdk ) based on @AdrianLxM  work. I believe it can be integrated into xDrip as a module too without modification maybe except some gradle stuff. This module contains code for parsing all kinds of data. Then it would be super easy exchange data between xdrip and AAPS and it give you the possibility to start using NS api v3 in xDrip as well
4) drop and use data only as decribed above?

**Reply by jamorham** [reply]
1) This sounds good
2) It is only the receiving of glucose values which is dependent on follower mode afaik, all other treatment type data should just work
3) This could be useful. I haven't examined it in any detail to know exactly how it could be used.
4) Typically we would maintain some legacy compatibility but if it is no longer needed then it can be removed. I think it is disabled by default anyway.

**Comment by MilosKozak** [comment]
Great. So first step is integrate ns-sdk to xDrip+ and let me know the intent. Maybe would be good to change it because timestamps are different to v1 code. Is there anyone who can work on it?
If you make it compile I can provide hints regarding usage

**Comment by paravoid** [comment]
(It's great to see this level of collaboration between the two projects :) Thank you both so much!)

If I may plug an older issue of my own that may be relevant to this discussion: https://github.com/NightscoutFoundation/xDrip/discussions/2506. The short version is that currently no (functional) way exists for `Sensor Start`/`Sensor Stop` events to find their way into Nightscout when one uses both projects, as 1) xDrip considers these events as "treatments", 2) these events are not broadcasted to AAPS 3) and when using both applications, AAPS should exclusively handle treatment uploads to Nightscout (hence the AAPS documentation requests to Disable "Upload treatments" in xDrip).

**Comment by MilosKozak** [comment]
@jamorham code is in dev of AAPS

**Reply by jamorham** [reply]
@MilosKozak which code? the ns-sdk or something else?

**Comment by jamorham** [comment]
Here is some example code that stimulates the test receiver on xDrip side:

```
        val gson = Gson()
        val nsds = NSDeviceStatus(
            "test", "test id", tsl(), tsl(), "test created", tsl(), 100, true, "test device", null,
            NSDeviceStatus.Pump("test clock", 99.9, null, null, null, null), null, null
        )
        val intent = Intent()
        intent.`package` = "com.eveningoutpost.dexdrip"
        intent.action = "info.nightscout.client.NS_BRIDGE"
        intent.putExtra("devicestatus", gson.toJson(nsds))
        sendBroadcast(intent)
```

**Reply by MilosKozak** [reply]
yes ns-sdk module
and this is data sending code
https://github.com/nightscout/AndroidAPS/blob/dev/plugins/sync/src/main/java/info/nightscout/plugins/sync/xdrip/XdripPlugin.kt

so far I kept BG in old format but I'd change it one you put the code in place

**Reply by Navid200** [reply]
@MilosKozak 
Would you please open an issue and explain what needs to be done, which you have suggested in your post of Feb 21?
And if there has been any development since then, please update accordingly.
Please title it accordingly as you wish.  For example: 
**Add ns-sdk to xDrip for AAPS**

I ask everyone else to please stay out of the issue unless absolutely necessary to allow Milos and JamOrHam get minimal notifications.

**Comment by swissalpine** [comment]
> * [ ]  Topic 4)
>   Do we still need status line?

The status line is passed on to watches such as Garmin and Fitbit etc. via the web service.
For example, I process it with my Garmin Watchfaces to display IOB, COB and TBR on the watchfaces.
Therefore: Please leave it as it is! Or it would break this functionality.

<img width="265" alt="xDrip+ Watchface new venu sq" src="https://user-images.githubusercontent.com/12517693/220444086-b5cdc176-4be1-4419-986b-8550b16fd817.png">

**Comment by Navid200** [comment]
@MilosKozak What is the reason for this note on AAPS site (https://androidaps.readthedocs.io/en/latest/Configuration/Config-Builder.html)?

![Screenshot 2023-03-16 223807](https://user-images.githubusercontent.com/51497406/225798431-b0357357-785a-4d8c-8523-15424abf1050.png)

**Reply by MilosKozak** [reply]
It's outdated info. I removed it

**Reply by Navid200** [reply]
That note is still there!

**Reply by MilosKozak** [reply]
source has been updated. it appears on web on next wiki build

**Comment by vanelsberg** [comment]
@jamorham Really missing the option of syncing all data to xDrip+
Any outlook/roadmap on a possible fix?

**Comment by szantos** [comment]
When everything is prepared in both apps (xDrip and AAPS) the documentation should also be updated. There is only minor info about how to set up the data flow from AAPS to xDrip. In case of success I'd be ready to submit a PR for both docs with the solution.

Sadly this feature is not working for me (for months). Now I've tried again without success. What do I miss?

I have:
- Pixel 6 Pro
- Android 13 (2023-02)
- AAPS 3.2-dev 2023-03-31
- xDrip 2023-03-28

Config in AAPS:
- regular app permissions granted: notification, files, etc.
- not granted: phone, sms
- unlimited battery granted (no optimization)
- Config builder: xDrip+ and Data Broadcaster enabled
- preferences / xDrip+: everything enabled
- xDrip Tab: there are RUN, UPL, SENDING, RECIPIENT events
- xDrip Tab: Queue: n/a
- Ongoing notification active (silent, but I've tried default, too)

Config in xDrip+:
- regular app permissions granted: notification, files, etc.
- not granted: phone, sms, contacts, camera, call log
- unlimited battery granted (no optimization)
- accessibility granted
- device and app notification permission granted
- modify system settings permission granted
- Inter-app settings: accept glucose, treatments, external status on
- Inter-app settings: Broadcast service API: tried both on and off
- Inter-app settings: Broadcast locally and everything else is off
- Tried Hardware Data Source: 640G/Eversense and Companion App
- every time with restart collector or even kill app + clear cache

Can you help me, please?

**Reply by szantos** [reply]
> You should set xDrip+ hardware data source to **xDrip+ Sync Follower

Thanks! That did it. **"xDrip+ Sync Follower / AndroidAPS"** would be a more suitable name for this option, what do you think @Navid200 ?

I hope, treatments will arrive, too. I like to see the sum of my carbs and insulin for the current day. Sadly AAPS can't display these figures live on its main screen or widget.

**Comment by Navid200** [comment]
I respectfully ask everyone to stay out of this thread.
This thread was created for Milos and JamOrHam to communicate.  They both get notifications for every new post in this thread.


If you need help, if you have any suggestions, if you want to tag me, if you want to tag either JamOrHam or Milos, please create a new thread.
